route_guide/route_guide_server.cc

```diff
@@ -55,166 +66,259 @@ float GetDistance(const Point& start, const Point& end) {
-        f.location().longitude() == point.longitude()) {
+    if (google::protobuf::util::MessageDifferencer::Equals(
+            f.location(),
+            point)) {
       return f.name();
     }
   }
   return "";
 }
 
+
+// The RouteGuideImpl class demonstrates server-side use of stout-grpc,
+// which is a higher-level interface on top of gRPC while still being
+// asynchronous. It uses stout-eventuals
+// (https://github.com/3rdparty/stout-eventuals) to make composing async code
+// easier.
+//
+// This class contains 4 methods: ServeGetFeature (unary rpc),
+// ServeListFeatures (server streaming), ServeRecordRoute (client streaming),
+// and ServeRouteChat (bidirectional streaming).
 class RouteGuideImpl final {
  public:
+  // Populate feature_list_ vector with addresses from database
+  // (route_guide_db.json).
   explicit RouteGuideImpl(const std::string& db) {
     routeguide::ParseDb(db, &feature_list_);
   }
 
-  auto Serve(Server* server) {
-    auto status = server->Serve<RouteGuide, Point, Feature>(
-        "GetFeature",
-        [this](auto* call, auto&& point) {
-          Feature feature;
-          feature.set_name(GetFeatureName(*point, feature_list_));
-          feature.mutable_location()->CopyFrom(*point);
-          call->WriteAndFinish(feature, Status::OK);
-        },
-        [](auto*, bool) {});
-
-    if (!status.ok()) {
-      return status;
-    }
+  // ServeGetFeature serves unary (one-request, one-response) rpc calls.
+  // The client sends the server a request containing a Point (latitude,
+  // longitude). If the server knows the address that corresponds to the
+  // Point's coordinates it will response with that address.
+  //
+  // The `Accept<...>()` method (stout/grpc/server.h) gets the name of the
+  // RPC function, and creates an "Eventual" that will produce a stream of
+  // connections. Eventuals can be composed with other Eventuals using
+  // the "|" operator, and here we compose `Accept` with a subsequent
+  // `Map` to handle each connection in the stream with a `Server::Handler()`.
+  // We end the chain of Eventuals processing each connection by composing
+  // with `Loop`, which takes us back to the top of the streaming chain to
+  // process the next connection.
+  auto ServeGetFeature(Server* server) {
+    return server->Accept<
+               RouteGuide, // Service name.
+               Point, // Request from client.
+               Feature>( // Response from the server.
+               "GetFeature")
+        | Map(Then([this](auto&& context) {
+             // The Server::Handler() can be composed with other tasks using
+             // the "|" operator.
+             // Its .body() callback is invoked every time the client sends a
+             // request to the server.
+             return Server::Handler(std::move(context))
+                 .body([this](auto& call, auto&& point) {
+                   // Check if the request is valid.
+                   if (!point) {
```

Comment:
As with the comments below about type-safe checks:
```suggestion
                   if (point == nullptr) {
```